subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,121 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	/* The number of struct bt_ad elements. */
+	size_t ad_len;
+
+	/* The current index in the array of struct bt_ad elements */
+	size_t ad_index;
+	/* The current index in the array of ad.data elements */
+	size_t data_index;
+
+	/* Current LTV offset contains the offset in the AD structure.
+	 * The length and type are included in this offset.
+	 */
+	uint16_t current_ltv_offset;
+};
+
+static void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	(void)memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+/** @brief Returns true if the current stream is empty.
+ *
+ * @param stream AD stream, @ref ad_stream_new
+ *
+ *
+ * @returns true if the stream is now empty.
+ */
+static bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
```

Comment:
Why was this changed back to index/len comparison, instead of a `NULL` check?

Reply:
When refactoring it caused `ad_stream_current_ltv_update()` to look a bit ugly. Discussed with @alwa-nordic

Reply:
Then please add a comment for this function. 

While it makes sense for me right now, it was confusing to me at first, so will likely be that for me again later, or for others. 

Something along the line of `if ad_index == ad_len, then we are past the last element in the ad array`